File: scraper/src/github.py
#!/usr/bin/env python3

import argparse
import json
import logging
import re
from datetime import datetime, timedelta
from os import getenv
from pathlib import Path
from urllib.parse import parse_qsl, urlparse
from zoneinfo import ZoneInfo

import requests

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] - %(message)s",
)

logger = logging.getLogger(__name__)


user_blacklist = {
    "dependabot",
    "snyk-bot",
    "codecov-commenter",
}


def serializer(obj):
    return obj.timestamp() if isinstance(obj, datetime) else repr(obj)


class GitHubScraper:
    def __init__(
        self, org, token, data_dir, date, days_back=1, log_level=logging.INFO
    ):
        self.log = logging.getLogger("GitHubScraper")
        self.log.setLevel(log_level)
        self.org = org
        self.token = token
        self.start_date = (date - timedelta(days=int(days_back))).date()
        self.end_date = date.date()
        self.data = {}
        self.data_dir = data_dir
        self.headers = {
            "Authorization": f"token {self.token}",
            # https://docs.github.com/en/rest/overview/media-types
            "Accept": "application/vnd.github.v3.raw+json",
        }
        self.log.debug(
            f"Fetching events for {self.org} from {self.start_date} to {self.end_date}"
        )

    def append(self, user, event):
        self.log.debug(f"Appending event for {user}")
        try:
            self.data[user]["activity"].append(event)
            if event["time"] > self.data[user]["last_updated"]:
                self.data[user]["last_updated"] = event["time"]
        except KeyError:
            self.log.debug(f"Adding new user {user}")
            self.data[user] = {
                "last_updated": event["time"],
                "activity": [event],
                "open_prs": [],
                "authored_issue_and_pr":[],
            }

    def parse_event(self, event, event_time):
        user = event["actor"]["login"]
        try:
            _user = event["payload"]["pull_request"]["user"]["login"]
        except KeyError:
            _user = user
        if _user.endswith("[bot]") or _user in user_blacklist:
            self.log.debug(f"Skipping blacklisted user {_user}")
            return

        self.log.debug(f"Parsing event for {user}")
        self.log.debug(f"event_id: {event['id']}")
        # if event["type"] == "cross-referenced":
        #     print(event)
        if event["type"] == "IssueCommentEvent":
            if event["payload"]["action"] in ("created",):
                self.append(
                    event["actor"]["login"],
                    {
                        "type": f'comment_{event["payload"]["action"]}',
                        "title": f'{event["repo"]["name"]}#{event["payload"]["issue"]["number"]}',
                        "time": event_time,
                        "link": event["payload"]["comment"]["html_url"],
                        "text": event["payload"]["comment"]["body"],
                    },
                )

        elif event["type"] == "IssuesEvent":
            if event["payload"]["action"] in (
                "opened",
                "assigned",
                "closed",
            ):
                self.append(
                    user,
                    {
                        "type": f'issue_{event["payload"]["action"]}',
                        "title": f'{event["repo"]["name"]}#{event["payload"]["issue"]["number"]}',
                        "time": event_time,
                        "link": event["payload"]["issue"]["html_url"],
                        "text": event["payload"]["issue"]["title"],
                    },
                )

        elif event["type"] == "PullRequestEvent":
            if event["payload"]["action"] == "opened":
                self.append(
                    user,
                    {
                        "type": f'pr_{event["payload"]["action"]}',
                        "title": f'{event["repo"]["name"]}#{event["payload"]["pull_request"]["number"]}',
                        "time": event_time,
                        "link": event["payload"]["pull_request"]["html_url"],
                        "text": event["payload"]["pull_request"]["title"],
                    },
                )

            elif (
                event["payload"]["action"] == "closed"
                and event["payload"]["pull_request"]["merged"]
            ):
                self.append(
                    event["payload"]["pull_request"]["user"]["login"],
                    {
                        "type": "pr_merged",
                        "title": f'{event["repo"]["name"]}#{event["payload"]["pull_request"]["number"]}',
                        "time": event_time,
                        "link": event["payload"]["pull_request"]["html_url"],
                        "text": event["payload"]["pull_request"]["title"],
                    },
                )

                self.add_collaborations(event, event_time)

        elif event["type"] == "PullRequestReviewEvent":
            self.append(
                user,
                {
                    "type": "pr_reviewed",
                    "time": event_time,
                    "title": f'{event["repo"]["name"]}#{event["payload"]["pull_request"]["number"]}',
                    "link": event["payload"]["review"]["html_url"],
                    "text": event["payload"]["pull_request"]["title"],
                },
            )
    
    def add_collaborations(self, event, event_time):
        collaborators = set()
        
        commits = requests.get(
            event['payload']['pull_request']['commits_url'],
            headers=self.headers,
        ).json()

        for commit in commits:
            collaborators.add(commit['author']['login'])

            co_authors = re.findall('Co-authored-by: (.+) <(.+)>', commit['commit']['message'])
            if co_authors:
                for (name, email) in co_authors:
                    users = requests.get(
                        'https://api.github.com/search/users',
                        params={'q': email},
                        headers=self.headers
                    ).json()

                    if users['total_count'] > 0:
                        collaborators.add(users['items'][0]['login'])
                        continue

                    users = requests.get(
                        'https://api.github.com/search/users',
                        params={'q': name},
                        headers=self.headers
                    ).json()

                    if users['total_count'] == 1:
                        collaborators.add(users['items'][0]['login'])
            
        if len(collaborators) > 1:
            for user in collaborators:
                others = collaborators.copy()
                others.remove(user)
                self.append(
                    user,
                    {
                        "type": "pr_collaborated",
                        "title": f'{event["repo"]["name"]}#{event["payload"]["pull_request"]["number"]}',
                        "time": event_time,
                        "link": event["payload"]["pull_request"]["html_url"],
                        "text": event["payload"]["pull_request"]["title"],
                        "collaborated_with": list(others)
                    },
                )


    def fetch_events(self, page=1):
        self.log.info(f"Fetching events page:{page}")
        resp = requests.get(
            f"https://api.github.com/orgs/{self.org}/events?per_page=100&page={page}",
            headers=self.headers,
        )
        if resp.status_code == 422:
            self.log.warning("Last page reached")  # stop pagination
            return self.data
        resp.raise_for_status()
        events = resp.json()

        events_count = 0
        for event in events:
            event_time = datetime.strptime(
                event["created_at"], "%Y-%m-%dT%H:%M:%S%z")

            if event_time.date() > self.end_date:
                continue
            elif event_time.date() <= self.start_date:
                return self.data
            self.parse_event(event, event_time)
            events_count += 1
        self.log.info(f"Fetched {events_count} events")

        if has_next := resp.links.get("next", {}).get("url"):
            next_page = dict(parse_qsl(urlparse(has_next).query)).get("page", 99)
            return self.fetch_events(int(next_page))
        return self.data

    def fetch_open_pulls(self, user):
        self.log.debug(f"Fetching open pull requests for {user}")
        resp = requests.get(
            f"https://api.github.com/search/issues?q=is:pr+is:open+org:{self.org}+author:{user}",
            headers=self.headers,
        )
        if resp.status_code == 422:
            self.log.warning("Last page reached")  # stop pagination
            return self.data
        resp.raise_for_status()
        pulls = resp.json()["items"]

        for pr in pulls:
            today = datetime.now()
            pr_last_updated = datetime.strptime(pr['updated_at'], "%Y-%m-%dT%H:%M:%SZ")

            self.data[user]["open_prs"].append(
                {
                    "link": pr["html_url"],
                    "title": pr["title"],
                    "stale_for": (today.replace(tzinfo=None) - pr_last_updated.replace(tzinfo=None)).days,
                    "labels": [label["name"] for label in pr["labels"]],
                }
            )
        self.log.debug(f"Fetched {len(pulls)} open pull requests for {user}")

        return self.data

    def resolve_autonomy_responsibility(self, event, user):
        if event["event"] == "cross-referenced" and event["source"]["type"] == "issue":
            return event["source"]["issue"]["user"]["login"] == user
        return False

    def fetch_merge_events(self, user):
        self.log.debug(f"Merge events for {user}")
        # check for those issues which are closed today (reduce number of issues to be calculated)
        resp = requests.get(
            f"https://api.github.com/search/issues?q=is:issue+is:closed+org:{self.org}+author:{user}",
            headers=self.headers,
        )
        if resp.status_code == 422:
            self.log.warning("Last page reached")  # stop pagination
            return self.data
        resp.raise_for_status()
        issues = resp.json()["items"]
        merged_prs = []
        for issue in issues:
            timeline_events = requests.get(
                issue["timeline_url"],
                headers=self.headers,
            )
            if timeline_events.status_code != 200:
                return
            timeline_events.raise_for_status()
            events = timeline_events.json()
            for event in events:
                if self.resolve_autonomy_responsibility(event, user):
                    if 'pull_request' in event["source"]["issue"]:
                        pr = event["source"]["issue"]["pull_request"]
                        if pr["merged_at"]:
                            merged_prs.append(
                                {
                                    "issue_link": issue["html_url"],
                                    "pr_link": pr["html_url"]
                                }
                            )
        for pr in merged_prs:
            self.data[user]["authored_issue_and_pr"].append(pr)
        self.log.debug(f"Fetched {len(merged_prs)} merged pull requests and issues for {user}")
        return self.data

    def scrape(self):
        self.log.info(f"Scraping {self.org}")
        self.fetch_events(1)
        self.log.info(f"Scraping open pull requests for {len(self.data)} users")
        for user in self.data.keys():
            try:
                self.fetch_merge_events(user)
            except Exception as e:
                self.log.error(f"Error fetching merge events for {user}: {e}")
            try:
                self.fetch_open_pulls(user)
            except Exception as e:
                self.log.error(f"Error fetching open pulls for {user}: {e}")
        self.log.info(f"Scraped {self.org}")
        return self.data

    def load_user_data(self, user):
        file = self.data_dir / f"{user}.json"
        self.log.debug(f"Loading user data from {file}")
        try:
            with file.open() as f:
                return json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            self.log.debug(f"User data not found for {user}")
            return {"activity": []}

    def save_user_data(self, user, data):
        file = self.data_dir / f"{user}.json"
        self.log.debug(f"Saving user data to {file}")
        with file.open("w") as f:
            json.dump(data, f, indent=2, default=serializer)

    def merge_data(self):
        self.log.info("Updating data")
        for user in self.data.keys():
            self.log.debug(f"Merging user data for {user}")
            old_data = self.load_user_data(user)
            data = self.data.get(user)
            data["activity"].extend(old_data["activity"])
            self.save_user_data(user, data)
        self.log.info("Updated data")

    def run(self):
        self.scrape()
        self.merge_data()
        self.log.info("Done")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("org_name", help="GitHub organization")
    parser.add_argument("data_dir", help="Directory to store user data")
    parser.add_argument(
        "-d",
        "--date",
        required=False,
        help="Date to start scraping from (YYYY-MM-DD)",
        default=None,
    )
    parser.add_argument(
        "-n",
        "--num-days",
        required=False,
        help="Number of days to scrape",
        default=1,
    )
    parser.add_argument(
        "-l",
        "--log-level",
        required=False,
        dest="loglevel",
        choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"],
        help="Set the logging level",
        default="INFO",
    )
    args = parser.parse_args()

    logger.setLevel(getattr(logging, args.loglevel))

    token = getenv("GITHUB_TOKEN")
    if not token:
        logger.error("GITHUB_TOKEN not found in environment")
        exit(1)

    if args.date is None:
        date = datetime.now(tz=ZoneInfo("UTC")) - timedelta(days=1)
    else:
        date = datetime.strptime(date, "%Y-%m-%d").replace(tzinfo=ZoneInfo("UTC"))

    scraper = GitHubScraper(
        args.org_name,
        token,
        Path(args.data_dir),
        date,
        days_back=args.num_days,
        log_level=getattr(logging, args.loglevel),
    )
    scraper.run()


if __name__ == "__main__":
    main()

File: scraper/src/github.ts
import { formatISO, parseISO, startOfDay, subDays } from "date-fns";
import fs from "fs";
import path from "path";
import { Activity, ActivityData, ProcessData, Action } from "../../lib/types";
import { PullRequestEvent, IGitHubEvent } from "../../lib/gh_events";
import { Octokit } from "octokit";

let processedData: ProcessData = {};

const GITHUB_TOKEN = process.env.GITHUB_TOKEN;
if (!GITHUB_TOKEN) {
  console.error("GITHUB_TOKEN not found in environment");
  process.exit(1);
}
const octokit = new Octokit({
  auth: GITHUB_TOKEN,
});

const fetchEvents = async (org: string, startDate: Date, endDate: Date) => {
  const events = await octokit.paginate(
    "GET /orgs/{org}/events",
    {
      org: org,
      per_page: 1000,
    },
    (response: { data: IGitHubEvent[] }) => {
      return response.data;
    },
  );

  let eventsCount: number = 0;
  let filteredEvents = [];
  for (const event of events) {
    const eventTime: Date = new Date(event.created_at ?? 0);

    if (eventTime > endDate) {
      continue;
    } else if (eventTime <= startDate) {
      return filteredEvents;
    }
    const isBlacklisted: boolean = [
      "dependabot",
      "snyk-bot",
      "codecov-commenter",
      "github-actions[bot]",
    ].includes(event.actor.login);
    const isRequiredEventType: boolean = [
      "IssueCommentEvent",
      "IssuesEvent",
      "PullRequestEvent",
      "PullRequestReviewEvent",
    ].includes(event.type ?? "");
    console.log(isRequiredEventType);
    if (!isBlacklisted && isRequiredEventType) {
      console.log(event.type);
      filteredEvents.push(event);
    }
    eventsCount++;
  }
  console.log("Fetched " + { eventsCount } + " events");

  return filteredEvents;
};
function appendEvent(user: string, event: Activity) {
  console.log(`Appending event for ${user}`);
  if (!processedData[user]) {
    console.log(`Creating new user data for ${user}`);
    processedData[user] = {
      last_updated: event.time,
      activity: [event],
      open_prs: [],
      authored_issue_and_pr: [],
    };
  } else {
    processedData[user]["activity"].push(event);
    if (event["time"] > (processedData[user]["last_updated"] ?? 0)) {
      processedData[user]["last_updated"] = event["time"];
    }
  }
}
const userBlacklist = new Set(["dependabot", "snyk-bot", "codecov-commenter"]);

const isBlacklisted = (login: string): boolean => {
  return login.includes("[bot]") || userBlacklist.has(login);
};
function parseISODate(isoDate: Date) {
  return new Date(isoDate);
}
async function calculateTurnaroundTime(event: PullRequestEvent) {
  const user: string = event.payload.pull_request.user.login;
  const mergedAt: Date = parseISODate(event.payload.pull_request.merged_at);
  const createdAt: Date = parseISODate(event.payload.pull_request.created_at);

  const linkedIssues: [string, string][] = [];
  const body = event.payload.pull_request.body || "";
  const regex =
    /(fix|fixes|fixed|close|closes|closed|resolve|resolves|resolved) ([\w\/.-]*)(#\d+)/gi;
  let match: RegExpExecArray | null;

  while ((match = regex.exec(body)) !== null) {
    linkedIssues.push([match[2], match[3]]);
  }

  const prTimelineResponse = await octokit.request(
    `GET ${event.payload?.pull_request?.issue_url}/timeline`,
  );

  const prTimeline = prTimelineResponse.data;

  prTimeline.forEach((action: Action) => {
    if (
      action.event === "cross-referenced" &&
      action.source.type === "issue" &&
      !action.source.issue.pull_request
    ) {
      linkedIssues.push([
        action.source.issue.repository.full_name,
        `#${action.source.issue.number}`,
      ]);
    }

    if (action.event === "connected") {
      // TODO: currently there is no way to get the issue number from the timeline, handle this case while moving to graphql
    }
  });
  const uniqueLinkedIssues: [string, string][] = Array.from(
    new Set(linkedIssues.map((issue) => JSON.stringify(issue))),
  ).map((item) => JSON.parse(item) as [string, string]);
  const assignedAts: { issue: string; time: Date }[] = [];

  for (const [org_repo, issue] of uniqueLinkedIssues) {
    const org = org_repo.split("/")[0] || event.repo.name.split("/")[0];
    const repo = org_repo.split("/")[-1] || event.repo.name.split("/")[1];
    const issueNumber = parseInt(issue.split("#")[1]);

    const issueTimelineResponse = await octokit.request(
      "GET /repos/{owner}/{repo}/issues/{issue_number}/timeline",
      {
        owner: org,
        repo: repo,
        issue_number: issueNumber,
      },
    );

    const issueTimeline = issueTimelineResponse.data;
    issueTimeline.forEach((action: Action) => {
      if (action.event === "assigned" && action.assignee.login === user) {
        assignedAts.push({
          issue: `${org}/${repo}#${issueNumber}`,
          time: parseISODate(action.created_at),
        });
      }

      if (action.event === "unassigned" && action.assignee.login === user) {
        assignedAts.pop();
      }
    });
  }

  const assignedAt: Date | null =
    assignedAts.length === 0
      ? null
      : assignedAts.reduce((min, current) =>
          current.time < min.time ? current : min,
        ).time;
  const turnaroundTime =
    (mergedAt.getTime() - (assignedAt || createdAt.getTime()).valueOf()) / 1000;
  return turnaroundTime;
}
async function addCollaborations(event: PullRequestEvent, eventTime: Date) {
  let nameUserCache: { [key: string]: string } = {};
  let emailUserCache: { [key: string]: string } = {};
  const collaborators: Set<string> = new Set();

  const url: string | undefined = event.payload.pull_request?.commits_url;

  const response = await octokit.request("GET " + url);
  const commits = response.data;
  for (const commit of commits) {
    let authorLogin = commit.author && commit.author.login;
    if (!authorLogin) {
      authorLogin = commit.commit.author.name;
    }

    if (isBlacklisted(authorLogin)) {
      continue;
    }

    collaborators.add(authorLogin);

    const coAuthors = commit.commit.message.match(
      /Co-authored-by: (.+) <(.+)>/,
    );
    if (coAuthors) {
      for (const [name, email] of coAuthors) {
        if (isBlacklisted(name)) {
          continue;
        }

        if (name in nameUserCache) {
          collaborators.add(nameUserCache[name]);
          continue;
        }

        if (email in emailUserCache) {
          collaborators.add(emailUserCache[email]);
          continue;
        }

        try {
          const usersByEmail = await octokit.request("GET /search/users", {
            q: email,
          });

          if (usersByEmail.data.total_count > 0) {
            const login = usersByEmail.data.items[0].login;
            emailUserCache[email] = login;
            collaborators.add(login);
            continue;
          }
          const usersByName = await octokit.request("GET /search/users", {
            q: name,
          });

          if (usersByName.data.total_count === 1) {
            const login = usersByName.data.items[0].login;
            nameUserCache[name] = login;
            collaborators.add(login);
          }
        } catch (e) {
          console.error(
            `Error fetching co-authors for commit ${commit} - ${name} <${email}>: ${e}`,
          );
        }
      }
    }
  }

  if (collaborators.size > 1) {
    const collaboratorArray = Array.from(collaborators); // Convert Set to Array
    for (const user of collaboratorArray) {
      const others = new Set(collaborators);
      const othersArray = Array.from(others);

      others.delete(user);
      appendEvent(user, {
        type: "pr_collaborated",
        title: `${event.repo.name}#${event.payload.pull_request.number}`,
        time: eventTime.toISOString(),
        link: event.payload.pull_request.html_url,
        text: event.payload.pull_request.title,
        collaborated_with: [...othersArray],
      });
    }
  }
}
async function resolve_autonomy_responsibility(event: Action, user: string) {
  if (event.event === "cross-referenced" && event.source.type === "issue") {
    return event.source.issue.user.login === user;
  }
  return false;
}
const fetch_merge_events = async (user: string, org: string) => {
  console.log("Merge events for : ", user);

  // Fetching closed issues authored by the user
  const { data: issues } = await octokit.request("GET /search/issues", {
    q: `is:issue is:closed org:${org} author:${user}`,
  });

  let merged_prs = [];

  for (const issue of issues.items) {
    const { data: timeline_events } = await octokit.request(
      "GET " + issue.timeline_url,
    );

    for (const event of timeline_events) {
      if (await resolve_autonomy_responsibility(event, user)) {
        const pull_request = event.source.issue.pull_request;
        if (pull_request && pull_request.merged_at) {
          merged_prs.push({
            issue_link: issue.html_url,
            pr_link: pull_request.html_url,
          });
        }
      }
    }
  }

  if (!processedData[user]) {
    processedData[user] = {
      authored_issue_and_pr: [],
      last_updated: "",
      activity: [],
      open_prs: [],
    };
  }

  for (const pr of merged_prs) {
    processedData[user].authored_issue_and_pr.push(pr);
  }

  return processedData;
};

const fetchOpenPulls = async (user: string, org: string) => {
  console.log(`Fetching open pull requests for ${user}`);
  const { data } = await octokit.request("GET /search/issues", {
    q: `is:pr is:open org:${org} author:${user}`,
  });

  type PullsData = (typeof data.items)[0];
  let pulls: PullsData[] = data.items;

  pulls.forEach((pr: PullsData) => {
    let today: Date = new Date();
    let prLastUpdated: Date = new Date(pr.updated_at);
    let staleFor: number = Math.floor(
      (today.getTime() - prLastUpdated.getTime()) / (1000 * 60 * 60 * 24),
    );

    if (!processedData[user]) {
      processedData[user] = {
        last_updated: "",
        activity: [],
        authored_issue_and_pr: [],
        open_prs: [],
      };
    }
    processedData[user].open_prs.push({
      link: pr.html_url,
      title: pr.title,
      stale_for: staleFor,
      labels: pr.labels.map((label: { name: string }) => label.name),
    });
  });

  console.log(`Fetched ${pulls.length} open pull requests for ${user}`);
  return processedData;
};
const parse_event = async (events: IGitHubEvent[]) => {
  for (const event of events) {
    const eventTime: Date = parseISO(event.created_at);
    const user: string = event.actor.login;
    if (isBlacklisted(user)) continue;

    console.log("Processing event for user: ", user);
    console.log("event_id : ", event.id);

    switch (event.type) {
      case "IssueCommentEvent":
        if (event.payload.action === "created") {
          appendEvent(user, {
            type: "comment_created",
            title: `${event.repo.name}#${event.payload.issue.number}`,
            time: eventTime.toISOString(),
            link: event.payload.comment.html_url,
            text: event.payload.comment.body,
          });
        }
        break;
      case "IssuesEvent":
        if (["opened", "assigned", "closed"].includes(event.payload.action)) {
          appendEvent(user, {
            type: `issue_${event.payload.action}`,
            title: `${event.repo.name}#${event.payload.issue?.number}`,
            time: eventTime.toISOString(),
            link: event.payload.issue.html_url,
            text: event.payload.issue.title,
          });
        }
        break;
      case "PullRequestEvent":
        if (event.payload.action === "opened") {
          appendEvent(user, {
            type: "pr_opened",
            title: `${event.repo.name}#${event.payload.pull_request.number}`,
            time: eventTime.toISOString(),
            link: event.payload.pull_request.html_url,
            text: event.payload.pull_request.title,
          });
        } else if (
          event.payload.action === "closed" &&
          event.payload.pull_request?.merged
        ) {
          const turnaroundTime: number = await calculateTurnaroundTime(event);
          appendEvent(user, {
            type: "pr_merged",
            title: `${event.repo.name}#${event.payload.pull_request.number}`,
            time: eventTime.toISOString(),
            link: event.payload.pull_request.html_url,
            text: event.payload.pull_request.title,
            turnaround_time: turnaroundTime,
          });
          await addCollaborations(event, eventTime);
        }
        break;
      case "PullRequestReviewEvent":
        appendEvent(user, {
          type: "pr_reviewed",
          time: eventTime.toISOString(),
          title: `${event.repo.name}#${event.payload.pull_request.number}`,
          link: event.payload.review.html_url,
          text: event.payload.pull_request.title,
        });
        break;
      default:
        break;
    }
  }
  return processedData;
};
function loadUserData(user: string, dataDir: string) {
  const file = path.join(dataDir, `${user}.json`);
  console.log(`Loading user data from ${file}`);

  try {
    const response = fs.readFileSync(file);
    const data: ActivityData = JSON.parse(response.toString());
    return data;
  } catch (error: any) {
    if (error.code === "ENOENT" || error.name === "SyntaxError") {
      console.log(`User data not found for ${user}`);
      return { activity: [] };
    } else {
      throw error; // rethrow unexpected errors
    }
  }
}
function saveUserData(
  user: string,
  data: ActivityData,
  dataDir: string,
  serializer: any,
) {
  const file = path.join(dataDir, `${user}.json`);
  console.log(`Saving user data to ${file}`);

  try {
    const jsonData = JSON.stringify(data, serializer, 2);
    fs.writeFileSync(file, jsonData);
  } catch (error: any) {
    console.error(`Failed to save user data for ${user}: ${error.message}`);
    throw error;
  }
}
const merged_data = async (dataDir: string) => {
  console.log("Updating data");
  fs.mkdirSync(dataDir, { recursive: true });

  for (let user in processedData) {
    if (processedData.hasOwnProperty(user)) {
      console.log(`Merging user data for ${user}`);
      let oldData = await loadUserData(user, dataDir);
      let userData = processedData[user];
      let newUniqueEvents = [];

      for (let event of userData.activity) {
        if (
          !oldData.activity.some(
            (oldEvent) => JSON.stringify(oldEvent) === JSON.stringify(event),
          )
        ) {
          newUniqueEvents.push(event);
        }
      }

      userData.activity = newUniqueEvents.concat(oldData.activity);
      saveUserData(user, userData, dataDir, null);
    }
  }
};

const scrapeGitHub = async (
  org: string,
  date: string,
  numDays: number = 1,
): Promise<void> => {
  const endDate: Date = startOfDay(parseISO(date));
  const startDate: Date = startOfDay(subDays(endDate, numDays));
  console.log(
    `Scraping GitHub data for ${org} from ${formatISO(startDate)} to ${formatISO(endDate)}`,
  );

  const events: IGitHubEvent[] = (await fetchEvents(
    org,
    startDate,
    endDate,
  )) as IGitHubEvent[];
  processedData = await parse_event(events);

  for (const user of Object.keys(processedData)) {
    try {
      await fetch_merge_events(user, org);
    } catch (e) {
      console.error(`Error fetching merge events for ${user}: ${e}`);
    }
    try {
      await fetchOpenPulls(user, org);
    } catch (e) {
      console.error(`Error fetching open pulls for ${user}: ${e}`);
    }
  }

  console.log("Scraping completed");
};

// Type Done and check done
const main = async () => {
  // Extract command line arguments (skip the first two default arguments)
  const args: string[] = process.argv.slice(2);

  // Destructure arguments with default values
  const [
    orgName,
    dataDir,
    date = formatISO(subDays(new Date(), 1), { representation: "date" }),
    numDays = 1,
  ] = args;

  if (!orgName || !dataDir) {
    console.error("Usage: node script.js <org> <dataDir> [date] [numDays]");
    process.exit(1);
  }

  await scrapeGitHub(orgName, date, Number(numDays));
  await merged_data(dataDir);
  console.log("Done");
};

main();
